### PR TITLE
🐛 Is638/ fixes dy-sidecar docker-compose process down command and timeout

### LINKED ISSUE #3172: Make docker_compose_down reliable
Change:
- this functions should raise an error if containers are still present by the time if finishes. 
- refactor to change order of operations (first run dowcker-compose down and then remove the pending stopped containers)
- add a regression test to ensure this happens and then fix

https://github.com/ITISFoundation/osparc-simcore/blob/3f88ad3d6340b40d616db8c4d0eee80d47da44ed/services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_compose_utils.py#L75-L99

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This PR addresses issues in the ``docker_compose_utils`` interface. Specifically:

- ♻️📝 Refactors ``docker_compose_utils`` interface
   -  Adds ``services/dynamic-sidecar/tests/unit/test_core_docker_compose_utils.py``
- 🐛 Fixes #3172 : ``rm`` is performed after properly ``down``
- 🐛 Timeout of ``async_command`` **did not terminate process**: *TL;DR*
  ```python
       await asyncio.wait_for( proc.communicate(), timeout=10) 
  ```
   will cancel the communications task but does NOT terminate the process.

Upgrades dynamic-sidecar to ``1.1.1``

## Related issue/s

- closes ITISFoundation/osparc-simcore#3172
- ITISFoundation/osparc-issues#638

## How to test

```cmd
cd services/dynamic-sidecar
make install-dev
pytest tests/unit/test_core_docker_compose_utils.py
pytest tests/unit/test_core_utils.py
```

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_compose_utils.py [hunk 1]
```diff
@@ -10,119 +9,150 @@
...
-    file_content: str,
+    compose_spec_yaml_content: str,
     command: str,
-    command_timeout: Optional[float],
+    terminate_process_on_timeout: Optional[int],
 ) -> CommandResult:
-    """The command which accepts {file_path} as an argument for string formatting"""
+    """The command which accepts {file_path} as an argument for string formatting
+
+    TODO: explain kill_on_timeout vs timeout in command
```
Comment (GitHK) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_compose_utils.py:18]: Has this been forgotten here?

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py [hunk 2]
```diff
@@ -107,29 +115,88 @@ async def write_to_tmp_file(file_contents: str) -> AsyncGenerator[Path, None]:
...
+        # NOTE: Sending SIGTERM to the process and awaits for termination
+        # i.e. giving the opportunity to the underying process to
+        # perform graceful shutdown (i.e. run shutdown events and cleanup tasks)
+        proc.terminate()
+        _close_transport(proc)
+
+        # There is a small risk that the process refused to shutdown
+        # when received a SIGTERM. For the moment, we will accept that risk
+        # and only check during development ...
+        #
```
Comment (GitHK) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py:162]: could you elaborate more when this happens?

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py [hunk 3]
```diff
@@ -107,29 +115,88 @@ async def write_to_tmp_file(file_contents: str) -> AsyncGenerator[Path, None]:
...
+            "Process %s timed out after %ss",
+            f"{command=!r}",
+            f"{timeout=}",
+        )
+        return CommandResult(
+            success=False,
+            message=f"Execution timed out after {timeout} secs",
+            command=f"{command}",
+            elapsed=time.time() - start,
+        )
```
Comment (GitHK) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py:176]: I'd guess there is no way to get the stdout at this point. Even if the command timed out?

File: services/dynamic-sidecar/tests/unit/test_core_utils.py [hunk 4]
```diff
@@ -17,8 +26,44 @@ def registry_with_auth(
...
+
+
+CURRENT_DIR = Path(sys.argv[0] if __name__ == "__main__" else __file__).resolve().parent
+
+
+@pytest.fixture
+def cmd(tmp_path: Path, sleep: int):
+    docker_compose = tmp_path / "docker_compose.yml"
+    docker_compose.write_text(
+        f"""\
```
Comment (GitHK) [inline services/dynamic-sidecar/tests/unit/test_core_utils.py:43]: maybe put the `f"""\` on the line above

File: services/dynamic-sidecar/tests/unit/test_core_docker_compose_utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    _print_result(r)
+    assert r.success, r.message
+
+    # restarts
+    r = await docker_compose_restart(
+        compose_spec_yaml,
+        settings,
+        10,
+    )
+    _print_result(r)
```
Comment (GitHK) [inline services/dynamic-sidecar/tests/unit/test_core_docker_compose_utils.py:87]: Note: this sometimes happens. Actually it only happens in one case. Could you make this test run with and without this step?

**Review by GitHK — COMMENTED** [review]
Looking nice. Please have a look at my comments below

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py (hunk 3 above)
Comment (pcrespov) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py:176]: It is indeed possible but will complicate much more the code. I propose to leave it as is and come back if it is really needed

**Review by GitHK — APPROVED** [review]
👍 great!

**Pull request merged**